application/src/states/track-loader.js

```diff
@@ -54,36 +54,84 @@ TrackLoaderState.prototype.preload = function()
 
 TrackLoaderState.prototype.create = function()
 {
-    var state = this;
-
     this.showTrackSelectorOffCanvas();
 
-    this.map = this.game.add.tilemap('track');
+    this.game.physics.startSystem(Phaser.Physics.P2JS);
+    this.game.physics.restitution = 0.8;
 
-    this.trackData.tilesets.forEach(function (tileset) {
-        state.map.addTilesetImage(tileset.name, 'tiles');
-    });
+    this.initTrack();
+
+    this.initPlayers();
+
+    this.game.physics.enable(this.car);
+
+    this.showLapCounter();
 
-    this.layer = this.map.createLayer('background');
+    this.game.camera.follow(this.car);
 
-    this.layer.resizeWorld();
+    this.game.scale.fullScreenScaleMode = Phaser.ScaleManager.SHOW_ALL;
+    this.game.input.onDown.add(this.toggleFullscreen, this);
 
     this.game.add.graphics();
+};
 
-    this.placeTrackMarkers();
+TrackLoaderState.prototype.initTrack = function()
+{
+    var backgroundLayer, dropsLayer, bodies, state = this;
 
-    this.game.physics.startSystem(Phaser.Physics.P2JS);
+    this.map = this.game.add.tilemap('track');
 
-    this.game.physics.restitution = 0.8;
+    this.trackData.tilesets.forEach(function (tileset) {
+        state.map.addTilesetImage(tileset.name, tileset.name);
+    });
 
-    this.collisionGroup = this.game.physics.p2.createCollisionGroup();
+    backgroundLayer = this.map.createLayer('background');
+    backgroundLayer.resizeWorld();
 
+    // Now that world size is set, we can create the main collision group
+    this.collisionGroup = this.game.physics.p2.createCollisionGroup();
     this.game.physics.p2.updateBoundsCollisionGroup();
 
-    this.car = this.carFactory.getNew(this.startingPoint[0], this.startingPoint[1], 'car');
-    this.game.world.addChild(this.car);
+    // Init pit layer
+    if (this.map.getLayerIndex('drops')) {
+        dropsLayer = this.map.createLayer('drops');
+        dropsLayer.visible = false;
+
+        // This code should work but it doesn't
+        this.map.setCollision(this.trackData.metaTileGid, true, dropsLayer);
+        bodies = this.game.physics.p2.convertTilemap(this.map, dropsLayer);
+        bodies.forEach(function (body) {
+            body.setCollisionGroup(state.collisionGroup);
+            body.collides(state.collisionGroup);
+            body.name = 'drop';
+        });
+
+        // Workaround for setTileIndexCallback, which for some reason isn't working
+        this.game.physics.p2.setPostBroadphaseCallback(function (body1, body2) {
+            var car, drop;
+            if (body1.name === 'drop') {
+                drop = body1;
+                car  = body2;
+            } else if (body2.name === 'drop') {
+                car  = body1;
+                drop = body2;
+            }
+            if (! state.car.falling) {
+                state.car.fall();
+            }
+        });
```

Comment:
Instead of using p2 physics bodies to detect whether the car is touching a pit tile, what do you think about doing an overlap check with the car object center? The p2 approach could work fine for pits, but for things like gravel and ice tiles we'll need to detect overlap instead of collision, and I think it would be easier if we detected contact with all specialty tiles the same way (as much as possible anyway).